pkg/healthcheck/healthcheck_output.go

```diff
@@ -114,95 +125,162 @@ func PrintChecksResult(wout io.Writer, output string, success bool, warning bool
-			//nolint:gosec
-			plugin := exec.Command(path, args...)
-			var stdout, stderr bytes.Buffer
-			plugin.Stdout = &stdout
-			plugin.Stderr = &stderr
-			plugin.Run()
-			extensionResults, err := parseJSONCheckOutput(stdout.Bytes())
-			spin.Stop()
-			if err != nil {
-				command := fmt.Sprintf("%s %s", path, strings.Join(args, " "))
-				if len(stderr.String()) > 0 {
-					err = errors.New(stderr.String())
-				} else {
-					err = fmt.Errorf("invalid extension check output from \"%s\" (JSON object expected):\n%s\n[%w]", command, stdout.String(), err)
-				}
-				results.Results = append(results.Results, CheckResult{
-					Category:    CategoryID(extensionCmd),
-					Description: fmt.Sprintf("Running: %s", command),
-					Err:         err,
-					HintURL:     HintBaseURL(version.Version) + "extensions",
-				})
-				success = false
+		plugin := utilsexec.Command(extension.path, args...)
+		var stdout, stderr bytes.Buffer
+		plugin.SetStdout(&stdout)
+		plugin.SetStderr(&stderr)
+		plugin.Run()
+		results, err := parseJSONCheckOutput(stdout.Bytes())
+		spin.Stop()
+		if err != nil {
+			success = false
+
+			command := fmt.Sprintf("%s %s", extension.path, strings.Join(args, " "))
+			if len(stderr.String()) > 0 {
+				err = errors.New(stderr.String())
 			} else {
-				results.Results = append(results.Results, extensionResults.Results...)
+				err = fmt.Errorf("invalid extension check output from \"%s\" (JSON object expected):\n%s\n[%w]", command, stdout.String(), err)
 			}
+			_, filename := filepath.Split(extension.path)
+			results = CheckResults{
+				Results: []CheckResult{
+					{
+						Category:    CategoryID(filename),
+						Description: fmt.Sprintf("Running: %s", command),
+						Err:         err,
+						HintURL:     HintBaseURL(version.Version) + "extensions",
+					},
+				},
+			}
+		}
+
+		extensionSuccess, extensionWarning := RunChecks(wout, werr, results, output)
+		if !extensionSuccess {
+			success = false
+		}
+		if extensionWarning {
+			warning = true
+		}
+	}
+
+	for _, m := range missing {
+		results := CheckResults{
+			Results: []CheckResult{
+				{
+					Category:    CategoryID(m),
+					Description: fmt.Sprintf("Linkerd extension command %s exists", m),
+					Err:         &exec.Error{Name: m, Err: exec.ErrNotFound},
+					HintURL:     HintBaseURL(version.Version) + "extensions",
+					Warning:     true,
+				},
+			},
 		}
 
-		var extensionSuccess bool
-		extensionSuccess, warning = RunChecks(wout, werr, results, output)
+		extensionSuccess, extensionWarning := RunChecks(wout, werr, results, output)
 		if !extensionSuccess {
 			success = false
 		}
+		if extensionWarning {
+			warning = true
+		}
 	}
 
 	return success, warning
 }
 
+// FindExtensions searches the path for all linkerd-* executables and returns a
+// slice of check commands, and a slice of missing checks.
+func FindExtensions(pathEnv string, glob Glob, exec utilsexec.Interface, nsLabels []string) ([]Extension, []string) {
```

Comment:
I'm finding this change a bit tough to follow because it mixes discovery of CLI extensions with handling of cluster extensions (`nsLabels`).  I think we can make this change a lot simpler and shrink this diff if we build it from functions which are more narrow and composable in what they do.

Specifically, if we make a functions like:
* `findCLIExtensionsOnPath() []string` // Returns all `linkerd-*` files on the PATH
* `findCLIChecks(files []string, exec utilsexec.Interface) []string // Filters the given executables to ones which are valid CLI checks

Then we can invoke these functions like:

```
extensionsSet := make(map[string]interface{})

namespaces, err := kubeAPI.GetAllNamespacesWithExtensionLabel(cmd.Context())
if err != nil {
    return false, false, err
}

for _, ns := range namespaces {
    extensionsSet[ns.Labels[k8s.LinkerdExtensionLabel]] = interface{}{}
}

cliExtensions := findCLIExtensionsOnPath()
cliChecks := findCLIChecks(cliExtensions, utilsexec.New())

for _, ext := range cliChecks {
    extensionsSet[strings.TrimPrefix(ext, "linkerd-")] = interface{}{}
}

extensions := []string{}
for ext := range extensionsSet {
    extensions = append(extensions, ext)
}
```

We could then invoke the existing `RunExtensionsChecks` functions without needing to change its behavior or signature at all.

I think there's a few advantages to doing things this way:
* Should be a much smaller diff, requiring no changes to `RunExtensionChecks` (except the removal of the extensions header)
* findCLIChecks is easier to test since it no longer requires passing in a glob function
* a large portion of the untestable OS behavior is isolated in the findCLIExtensionsOnPath function

Reply:
i can take a crack at breaking this down into smaller functions. a few notes:

> Specifically, if we make a functions like:
> - findCLIExtensionsOnPath() []string // Returns all linkerd-* files on the PATH

i think `getExtensionExecutables()` is already doing this?
- https://github.com/linkerd/linkerd2/pull/10588/files#diff-ad49e1a26f2606d0393d9088f4eccf5abbf7f62d8a3b58fa96fc9129e30a163dR516-R538

> findCLIChecks(files []string, exec utilsexec.Interface) []string // Filters the given executables to ones which are valid CLI checks

i think we're also pretty close here, via `isCLICheck()`?
- https://github.com/linkerd/linkerd2/pull/10588/files#diff-ad49e1a26f2606d0393d9088f4eccf5abbf7f62d8a3b58fa96fc9129e30a163dR228-R239

> Should be a much smaller diff, requiring no changes to RunExtensionChecks (except the removal of the extensions header)

i can give this a try, though i can't quite tell if leaving `RunExtensionChecks()` untouched makes sense. in `main` right now, `RunExtensionChecks()` calls `exec.LookPath()` on every candidate extension namespace, and also does special casing for jaeger/multicluster/viz. if we're already checking for `linkerd-*` executables on the path earlier in `getExtensionExecutables()`/`findCLIExtensionsOnPath()`, i'm not sure it makes sense to shell out again since we've already done the work to find every `linkerd-*` executable.

> a large portion of the untestable OS behavior is isolated in the findCLIExtensionsOnPath function

which parts are untestable? `FindExtensions()` takes a `Glob` and `utilsexec.Interface` to allow us to mock out filesystem searching and command execution, via `TestFindExtensions()`:
- https://github.com/linkerd/linkerd2/pull/10588/files#diff-ad49e1a26f2606d0393d9088f4eccf5abbf7f62d8a3b58fa96fc9129e30a163dR220
- https://github.com/linkerd/linkerd2/pull/10588/files#diff-e418f5d7f8c5588ed945b395f9722ddba164f7f8d66a81db45c9fece18529dc6R51

Reply:
By "untestable behavior" I mean the actual interactions with the OS that we would replace with mocks in a test.  By splitting all of that code into a function, we make it so that we need fewer mocks elsewhere.  And a function that only interacts with the OS doesn't need to be tested because if we did, we'd only be testing mocks.

To make that more concrete, I don't think it makes sense to unit test `findCLIExtensionsOnPath` because we'd be mocking out the entire behavior of that function in the test.  Therefore it doesn't need to accept glob as a parameter and we can eliminate the glob type.  Meanwhile, the actual logic of what we do with the results of the glob is testable in other functions.

Reply:
My instinct is that extra calls to `exec.LookPath()` shouldn't have a noticeable performance impact.  If we can simplify the code at the expense of some redundant calls to `exec.LookPath()`, I think it's worth it.

Reply:
i'm not concerned about the performance impact of extra calls `exec.LookPath()`, but rather i think the current approach yields simpler code, because we're searching for binaries on the filesystem in only one place.